File: src/main/java/com/github/dkoval/leetcode/challenge/UglyNumber2.java
package com.github.dkoval.leetcode.challenge;

import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Set;

/**
 * <a href="https://leetcode.com/problems/ugly-number-ii/">Ugly Number II</a>
 * <p>
 * An ugly number is a positive integer whose prime factors are limited to 2, 3, and 5.
 * <p>
 * Given an integer n, return the nth ugly number.
 * <p>
 * Constraints:
 * <p>
 * 1 <= n <= 1690
 */
public interface UglyNumber2 {

    int nthUglyNumber(int n);

    class UglyNumber2UsingMinHeap implements UglyNumber2 {

        private static final int[] FACTORS = {2, 3, 5};

        @Override
        public int nthUglyNumber(int n) {
            // generate the sequence of the first n numbers
            Queue<Long> minHeap = new PriorityQueue<>();
            Set<Long> seen = new HashSet<>();
            minHeap.offer(1L);
            seen.add(1L);
            int count = 0;
            while (true) {
                long curr = minHeap.poll();
                if (++count == n) {
                    return (int) curr;
                }

                for (int factor : FACTORS) {
                    long next = curr * factor;
                    if (!seen.contains(next)) {
                        minHeap.offer(next);
                        seen.add(next);
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/UglyNumber2.kt
package com.github.dkoval.leetcode.challenge

import kotlin.math.min

object UglyNumber2Smart : UglyNumber2 {

    override fun nthUglyNumber(n: Int): Int {
        val ugly = IntArray(n)
        ugly[0] = 1
        var idx2 = 0
        var idx3 = 0
        var idx5 = 0
        var i = 1
        while (i < n) {
            val next2 = ugly[idx2] * 2
            val next3 = ugly[idx3] * 3
            val next5 = ugly[idx5] * 5
            val next = min(min(next2, next3), next5)
            if (next == next2) idx2++
            if (next == next3) idx3++
            if (next == next5) idx5++
            ugly[i++] = next
        }
        return ugly[n - 1]
    }
}